src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +387,288 @@ public static void EnsureSocketsAreInitialized()
+            hints.ai_family = AddressFamily.Unspecified; // Gets all address families
+
+            SocketError errorCode =
+                (SocketError)Interop.Winsock.GetAddrInfoExW(name, null, 0 /* NS_ALL*/, IntPtr.Zero, ref hints, out context->Result, IntPtr.Zero, ref context->Overlapped, s_getAddrInfoExCallback, out context->CancelHandle);
+
+            if (errorCode != SocketError.IOPending)
+                ProcessResult(errorCode, context);
+        }
+
+        private static unsafe void GetAddressInfoExCallback([In] int error, [In] int bytes, [In] NativeOverlapped* overlapped)
+        {
+            // Can be casted directly to GetAddrInfoExContext* because the overlapped is its first field
+            GetAddrInfoExContext* context = (GetAddrInfoExContext*)overlapped;
+
+            ProcessResult((SocketError)error, context);
+        }
+
+        private static unsafe void ProcessResult(SocketError errorCode, GetAddrInfoExContext* context)
+        {
+            GetAddrInfoExState state = context->GetQueryState();
+
+            // 'state' may be null or 'SetCallbackStartedOrCanceled' may return false if
+            // the AppDomain started unloading before the callback was raised.
+            // In this case we don't need to free the context, it will be done in
+            // GetAddrInfoExState's finalizer.
+            if (state == null || !state.SetCallbackStartedOrCanceled())
+                return;
+
+            try
+            {
+                if (errorCode != SocketError.Success)
+                {
+                    state.CompleteAsyncResult(new SocketException((int)errorCode));
+                    return;
+                }
+
+                AddressInfoEx* result = context->Result;
+                string canonicalName = null;
+
+                List<IPAddress> addresses = new List<IPAddress>();
+
+                while (result != null)
+                {
+                    if (canonicalName == null && result->ai_canonname != IntPtr.Zero)
+                        canonicalName = Marshal.PtrToStringUni(result->ai_canonname);
+
+                    IPAddress ipAddress = null;
+                    var socketAddress = new ReadOnlySpan<byte>(result->ai_addr, result->ai_addrlen);
+
+                    if (result->ai_family == AddressFamily.InterNetwork)
+                    {
+                        if (socketAddress.Length == SocketAddressPal.IPv4AddressSize)
+                            ipAddress = CreateIPv4Address(socketAddress);
+                    }
+                    else if (SocketProtocolSupportPal.OSSupportsIPv6 && result->ai_family == AddressFamily.InterNetworkV6)
+                    {
+                        if (socketAddress.Length == SocketAddressPal.IPv6AddressSize)
+                            ipAddress = CreateIPv6Address(socketAddress);
+                    }
+
+                    if (ipAddress != null)
+                        addresses.Add(ipAddress);
```

Comment:
This addresses.Add can just be moved up to where the addresses are created, e.g. address.Add(CreateIPv4Address(socketAddress)).  The Create* methods will never return null, right?